cli/dcoscli/task/main.py

```diff
@@ -22,12 +24,92 @@ def main(argv):
         return 1
 
 
+def docopt_wrapper(usage, real_usage, **keywords):
+    """ A wrapper around the real docopt parser.
+    Redirects a failed exec command to /dev/null and prints the proper
+    real_usage message, instead of just the usage string from dcos-task-exec
+
+    :param usage: task id filter
+    :type usage: str
+    :param real_usage: If True, include completed tasks
+    :type real_usage: bool
+    :param keywords: If True, output json.  Otherwise, output a human
+                  readable table.
+    :type keywords: bool
+    :returns: process return code
+    """
+
+    try:
+        stdout = sys.stdout
+
+        with open(os.devnull, 'w') as nullfile:
+            sys.stdout = nullfile
+            arguments = docopt.docopt(usage, **keywords)
+            sys.stdout = stdout
+
+        return arguments
+
+    except docopt.DocoptExit:
+        sys.stdout = stdout
+        print(real_usage.strip(), file=sys.stderr)
+        sys.exit(1)
+
+    except SystemExit:
+        sys.stdout = stdout
+
+        if "argv" in keywords and any(h in ("-h", "--help")
+                                      for h in keywords["argv"]):
+            print(real_usage.strip())
+        elif "version" in keywords and any(v in ("--version")
+                                           for v in keywords["argv"]):
+            print(keywords["version"].strip())
+
+        sys.exit()
+
+
 @decorate_docopt_usage
 def _main(argv):
-    args = docopt.docopt(
-        default_doc("task"),
-        argv=argv,
-        version="dcos-task version {}".format(dcoscli.version))
+    """The main function for the 'task' subcommand"""
+
+    # We must special case the 'dcos task exec' subcommand in order to
+    # allow us to pass arguments to docopt in a more user-friendly
+    # manner. Specifically, we need to be able to to pass arguments
+    # beginning with "-" to the command we are trying to launch with
+    # 'exec' without docopt trying to match them to a named parameter
+    # for the 'dcos' command itself.
+    if len(argv) > 1 and argv[1] == "exec":
+        usage = """
+        Usage:
+            dcos-task-exec [--interactive --tty] <task> <cmd> [<args>...]
+
+        Options:
+            -i, --interactive
+            -t, --tty
+        """
+        args = docopt_wrapper(
+            usage,
+            default_doc("task"),
+            argv=argv[2:],
+            version="dcos-task version {}".format(dcoscli.version),
+            options_first=True)
+
+        # Unfortunately, we have to set the argument below to
+        # True/False because docopt does the same internally, and the
+        # CLI expects them to be set.
+        #
+        # TODO(slocke): If more commands are added to
+        # /dcos-cli/cli/dcoscli/data/help/task.txt they will need to
+        # be set to False here as well.
```

Comment:
I see. So run docopt twice -- once with the simplified usage doc (as is now) and once with the real usage doc (stripping the args beforehand so it doesn't blow up). Then merge the two?